lib/features/support/app_automate_driver.rb

```diff
@@ -0,0 +1,122 @@
+require 'appium_lib'
+require 'open3'
+require_relative './fast_selenium'
+
+# Wraps Appium::Driver to enable control of a BrowserStack app-automate session
+class AppAutomateDriver
+
+  # @!attribute [r] device_type
+  #   @return [String] The device, from the list of device capabilities, used for this test
+  attr_reader :device_type
+
+  # @!attribute [r] driver
+  #   @return [Appium::Driver, nil] The driver being used
+  attr_reader :driver
+
+  # @!attribute capabilities
+  #   @return [Hash] A hash of capabilities used by the driver
+  attr_accessor :capabilities
+
+  # The App upload uri for BrowserStack App Automate
+  APP_UPLOAD_URI = "https://api-cloud.browserstack.com/app-automate/upload"
+
+  # Creates the AppAutomateDriver
+  #
+  # An instance of this driver should be assigned to the $driver variable for the steps to work correctly.
+  # Adds an at_exit block to ensure a created driver is stopped
+  #
+  # @param username [String] the BrowserStack username
+  # @param access_key [String] the BrowserStack access key
+  # @param local_id [String] the identifier for the BrowserStackLocal tunnel
+  # @param locator [Symbol] the primary locator strategy Appium should use to find elements
+  def initialize(username, access_key, local_id, locator=:id)
+    @username = username
+    @access_key = access_key
+    @local_id = local_id
+    @capabilities = {
+      'browserstack.console': 'errors',
+      'browserstack.localIdentifier': local_id,
+      'browserstack.local': 'true',
+      'browserstack.networkLogs': 'true'
+    }
+    @locator = locator
+
+    at_exit do
+      @driver.quit unless @driver.nil?
+    end
+  end
+
+  # Initialises the BrowserStack app-automate connect by:
+  #   Uploading the app file
+  #   Starting the BrowserStackLocal tunnel
+  #   Creates and starts the Appium driver
+  #
+  # @param target_device [String] a key from the Devices array selecting which device capabilities to target
+  # @param app_location [String] the location of the test-app to upload
+  def start_driver(target_device, app_location)
+    @device_type = target_device
+    upload_app(app_location)
+    start_local_tunnel
+    @capabilities.merge! devices[target_device]
+    @driver = Appium::Driver.new({
+      'caps' => @capabilities,
+      'appium_lib' => {
+        :server_url => "http://#{@username}:#{@access_key}@hub-cloud.browserstack.com/wd/hub"
+      }
+    }, false).start_driver
+  end
+
+  # Checks for an element, waiting until it is present or the method times out
+  #
+  # @param element_id [String] the element to search for using the @locator strategy
+  # @param timeout [Integer] the maximum time to wait for an element to be present in seconds
+  def wait_for_element(element_id, timeout=15)
+    unless @driver.nil?
```

Comment:
As followup on https://github.com/bugsnag/maze-runner/pull/54#discussion_r284053696 , this design feels awkward because almost every public method checks `@driver.nil?` and each method has a test case that stubs `#nil?` (which doesn't feel right either).

An alternative approach might be to have something separate (whether it is another class or multiple classes or methods) perform the initialization (uploading the APK, starting the tunnel, creating an `Appium::Driver`), and then have this driver wrapper class's `#initialize` method take the `Appium::Driver` as an argument.

(However, I don't really see the benefit of having this `AppAutomateDriver` wrapper yet)

Reply:
I agree, but I can't think of a better solution right now.  Certain elements of this class will be abstracted out once the BrowserStack Browser-automation driver is extracted from Bugsnag-JS to here, and a proper design will likely be necessary to avoid too much complication.  Perhaps that'd be a good time to address these concerns?